Fix - Native editor buttons are incorrectly rendering on top of model metadata screen

kamilmielnik opened the pull request:
Fixes https://github.com/metabase/metabase/issues/30680

### Description

Removing an optimization fixes the issue.

### Demo

#### Before

https://github.com/metabase/metabase/assets/6830683/3ebe98ce-6681-4bda-9101-66ce280551ec


#### After

https://github.com/metabase/metabase/assets/6830683/22d2d134-4538-42e2-a13a-b448b67c7504

kamilmielnik committed:
Make NativeQueryEditor testable
- Mock just NativeQueryEditor.prototype.loadAceEditor instead of entire NativeQueryEditor

kamilmielnik committed:
Fix invalid propTypes

kamilmielnik committed:
Fix DatasetQueryEditor not re-rendering when props change
- remove custom arePropsEqual argument from React.memo() on NativeQueryEditor
- Fixes #30680

kamilmielnik committed:
Add a dirty initial NativeQueryEditor test setup

kamilmielnik committed:
Add DatasetQueryEditor tests

kamilmielnik committed:
Revert "Add a dirty initial NativeQueryEditor test setup"

This reverts commit 0f985db272a2e8f1be755bda07c3b642fdf90f83.

kamilmielnik committed:
Get rid of test errors

kamilmielnik committed:
Clean up DatasetQueryEditor test

kamilmielnik committed:
Add a test case for re-rendering of DatasetQueryEditor

kamilmielnik committed:
Make setup function more reusable

kamilmielnik committed:
Add 2 more test cases

kamilmielnik committed:
Bring back global NativeQueryEditor mock as many tests depend on it

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.jsx:
```diff
@@ -93,9 +93,4 @@ function DatasetQueryEditor({
 DatasetQueryEditor.propTypes = propTypes;
 
-export default React.memo(
-  DatasetQueryEditor,
```
If remove this we need to deal we the editor being always rendered another way. Looking at this code, could we conditionally render `DatasetQueryEditor` when `isEditingQuery` is `false`? https://github.com/metabase/metabase/blob/45f8791994d8f1af9d6e340cabbb529436707ea9/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx#L466

kulyk replied:
@ranquild re-mounting a notebook editor can be expensive if a query has a lot of clauses (see #12378), that's part of the reason why this `React.memo` was originally introduced

kamilmielnik replied:
### Measurements
I [measured](https://docs.google.com/spreadsheets/d/18HbARyTek1orJVx8jY79_j8GCWA4YTcP4-3OlF-kz64/edit#gid=0) the performance using [`React.Profiler`](https://react.dev/reference/react/Profiler), dev build, [this query](https://github.com/metabase/metabase/assets/6830683/86d717a7-19b0-4cad-a5fa-0391ae72360b) and viewport 1685 x 1335.

<details>
<summary>
Patch with code used to acquire these measurements
</summary>

Use `window.stats = {}` to clear measurements, use `window.printStats()` to log them.

```diff
diff --git a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
index d1abc185d6..660fdb2f8d 100644
--- a/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
+++ b/frontend/src/metabase/query_builder/components/DatasetEditor/DatasetEditor.jsx
@@ -1,4 +1,4 @@
-import React, { useEffect, useCallback, useMemo, useState } from "react";
+import React, { Profiler, useEffect, useCallback, useMemo, useState } from "react";
 import PropTypes from "prop-types";
 import { connect } from "react-redux";
 import { t } from "ttag";
@@ -463,28 +463,46 @@ function DatasetEditor(props) {
       <Root>
         <MainContainer>
           <QueryEditorContainer isResizable={isEditingQuery}>
-            <DatasetQueryEditor
-              {...props}
-              isActive={isEditingQuery}
-              height={editorHeight}
-              viewHeight={height}
-              onResizeStop={handleResize}
-            />
+            <Profiler
+              id="DatasetQueryEditor"
+              onRender={(id, phase, actualDuration) => {
+                window.stats = window.stats || {};
+                window.stats[id] = window.stats[id] || [];
+                window.stats[id].push({ phase, actualDuration });
+              }}
+            >
+              <DatasetQueryEditor
+                {...props}
+                isActive={isEditingQuery}
+                height={editorHeight}
+                viewHeight={height}
+                onResizeStop={handleResize}
+              />
+            </Profiler>
           </QueryEditorContainer>
           <TableContainer isSidebarOpen={!!sidebar}>
             <DebouncedFrame className="flex-full" enabled>
-              <QueryVisualization
-                {...props}
-                className="spread"
-                noHeader
-                queryBuilderMode="dataset"
-                isShowingDetailsOnlyColumns={datasetEditorTab === "metadata"}
-                hasMetadataPopovers={false}
-                handleVisualizationClick={handleTableElementClick}
-                tableHeaderHeight={isEditingMetadata && TABLE_HEADER_HEIGHT}
-                renderTableHeaderWrapper={renderTableHeaderWrapper}
-                scrollToColumn={focusedFieldIndex + scrollToColumnModifier}
-              />
+              <Profiler
+                id="QueryVisualization"
+                onRender={(id, phase, actualDuration) => {
+                  window.stats = window.stats || {};
+                  window.stats[id] = window.stats[id] || [];
+                  window.stats[id].push({ phase, actualDuration });
+                }}
+              >
+                <QueryVisualization
+                  {...props}
+                  className="spread"
+                  noHeader
+                  queryBuilderMode="dataset"
+                  isShowingDetailsOnlyColumns={datasetEditorTab === "metadata"}
+                  hasMetadataPopovers={false}
+                  handleVisualizationClick={handleTableElementClick}
+                  tableHeaderHeight={isEditingMetadata && TABLE_HEADER_HEIGHT}
+                  renderTableHeaderWrapper={renderTableHeaderWrapper}
+                  scrollToColumn={focusedFieldIndex + scrollToColumnModifier}
+                />
+              </Profiler>
             </DebouncedFrame>
             <TabHintToastContainer
               isVisible={isEditingMetadata && isTabHintVisible && !result.error}
diff --git a/frontend/src/metabase/query_builder/components/view/View.jsx b/frontend/src/metabase/query_builder/components/view/View.jsx
index 83d4a3708f..f8f5f01d59 100644
--- a/frontend/src/metabase/query_builder/components/view/View.jsx
+++ b/frontend/src/metabase/query_builder/components/view/View.jsx
@@ -1,7 +1,6 @@
 /* eslint-disable react/prop-types */
-import React from "react";
+import React, { Profiler } from "react";
 import { Motion, spring } from "react-motion";
-import _ from "underscore";
 import { t } from "ttag";
 
 import ExplicitSize from "metabase/components/ExplicitSize";
@@ -37,6 +36,13 @@ import ViewSidebar from "./ViewSidebar";
 import NewQuestionView from "./View/NewQuestionView";
 import QueryViewNotebook from "./View/QueryViewNotebook";
 
+window.printStats = () => {
+  Object.entries(window.stats).forEach(([key, value]) => {
+    const sum = value.reduce((result, { actualDuration }) => result + actualDuration, 0);
+    console.log(key, sum);
+  });
+}
+
 import {
   BorderedViewTitleHeader,
   NativeQueryEditorContainer,
@@ -464,7 +470,16 @@ class View extends React.Component {
     if (question.isDataset() && queryBuilderMode === "dataset") {
       return (
         <>
-          <DatasetEditor {...this.props} />
+          <Profiler
+            id="DatasetEditor"
+            onRender={(id, phase, actualDuration) => {
+              window.stats = window.stats || {};
+              window.stats[id] = window.stats[id] || [];
+              window.stats[id].push({ phase, actualDuration });
+            }}
+          >
+            <DatasetEditor {...this.props} />
+          </Profiler>
           <QueryModals {...this.props} />
         </>
       );
```

</details>

### Results
[Google doc](https://docs.google.com/spreadsheets/d/18HbARyTek1orJVx8jY79_j8GCWA4YTcP4-3OlF-kz64/edit#gid=0)

![image](https://github.com/metabase/metabase/assets/6830683/b817c289-cdef-468b-b833-af7a6832e450)

### Conclusion

Rendering `DatasetQueryEditor` conditionally in `DatasetEditor` is the best option.

kamilmielnik replied:
> ### Conclusion
> Rendering `DatasetQueryEditor` conditionally in `DatasetEditor` is the best option.

This will however break the height animation.

I feel like any attempt to make it work with JS will affect the performance. 
The animation makes sense only when there's not much to render. Otherwise (when performance is visibly lower) users won't see it anyway.

Can we just remove the animation?

#### Current, with animation

https://github.com/metabase/metabase/assets/6830683/9454a570-ed86-4e6e-99ea-e884523f4470

#### Conditional rendering, no animation

https://github.com/metabase/metabase/assets/6830683/bbe2aacd-ee92-439c-b184-ee2e30e57b84

kulyk replied:
Glad we can use conditional rendering here. Thanks for an excellent investigation, @kamilmielnik 🙏 

> This will however break the height animation.
>
> I feel like any attempt to make it work with JS will affect the performance.
The animation makes sense only when there's not much to render. Otherwise (when performance is visibly lower) users won't see it anyway.
> 
> Can we just remove the animation?

This is a tradeoff, but it makes sense to me. We had to disable the notebook slide-in/out animation for similar reasons.
Could you please double-check this with Maz and Bruno?

kamilmielnik replied:
> Could you please double-check this with Maz and Bruno?

The consensus is that performance is more important than the animation ([related Slack thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1685540949163939)).

kulyk commented:
@kamilmielnik @ranquild I tried out the branch with a pretty complex query (enough for the editor to become slow), and things are equally bad no matter if there's a `React.memo` or not. I'm worried unmounting the query editor on the metadata tab won't solve the issue — once you switch back to "Query", mounting the editor from scratch takes way longer.

We should address the notebook performance issue. FWIW I have a draft PR #29730 that partially solves it by migrating `ClauseStep` to lazy `TippyPopover`. Data and join steps still use the `DataSelector` wrapped with the tether popover and that's enough to keep the performance low

kamilmielnik commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,162 @@
+  return { query, question, rerender };
+};
+
+describe("DatasetQueryEditor", () => {
```
> I thinks this is a highly integration, browser rendering dependent case that we cannot test in Jest.

What do you mean by "cannot test in Jest"? I'm pretty sure I just did.

> Could you **instead** write an e2e test for this, with matching the reproduction steps as closely as possible? 

I do see an extra value in adding an e2e test for it (and I will) but why **instead**?

> The test should fail without your fix and pass with the fix applied.

For the record: that's exactly how the `re-renders DatasetQueryEditor when height is 0 and isActive prop changes` case works.

kamilmielnik committed:
Add an e2e test for #30680

kamilmielnik committed:
Optimize DatasetEditor by conditionally rendering DatasetQueryEditor only when it can be seen

kamilmielnik committed:
Remove redundant code

kamilmielnik commented:
@ranquild  @kulyk 

I have updated the PR:
- [I've preserved the optimization](https://github.com/metabase/metabase/pull/31142/commits/93dab29f4736cf8162d1047d6f897a634234173e) based on [fresh measurements](https://github.com/metabase/metabase/pull/31142#discussion_r1211352364)
- [I've added an e2e test for this issue](https://github.com/metabase/metabase/pull/31142/commits/b865c6328011291e90a8725ba86f037a2ea9fa74)

Please re-review.

The unintended consequence of this PR is that the [height animation on the query editor is removed](https://github.com/metabase/metabase/pull/31142#discussion_r1211367455).

kamilmielnik committed:
Merge branch 'master' into 30680-native-editor-buttons

kamilmielnik committed:
ESLint

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,161 @@
+  }),
+});
+
+const ROOT_COLLECTION = {
```
`= createMockCollection({ id: "root", ...})`. Please never cast invalid objects forcefully, this can become something very hard to migrate from.

kamilmielnik replied:
ee7c716bde4f5afb9bb9ae27a02faeb7d1c63e74

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,161 @@
+    }),
+  });
+  const metadata = getMetadata(storeInitialState);
+  const query = new NativeQuery(
```
I might be missing something but it seems that this can be simplified. You should be able to set `dataset_query` in `TEST_CARD`, and then `const question = question.query()` without constructing the query manually (which can be incorrect)

kamilmielnik replied:
ad1ed026b09d0a0761cfafcfff65ffe0496aea67

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,161 @@
+  );
+  const question = checkNotNull(metadata.question(card.id));
+
+  const { default: DatasetQueryEditor } = await import(
```
Why this is required?

kamilmielnik replied:
`NativeQueryEditor` is [globally mocked](https://github.com/metabase/metabase/blob/9ca4af5/frontend/test/register-visualizations.js#L10) but its actual implementation is needed in this test suite because we need to investigate its children.

We're actually testing `NativeQueryEditor` indirectly by using `DatasetQueryEditor` (which uses `NativeQueryEditor`), so the `NativeQueryEditor` has to be unmocked the moment we import `DatasetQueryEditor`.

Unmocking happens in `beforeEach`, so we can really only import the component during the unit test.

Should the import be at the beginning of this file, the [mock `NativeQueryEditor`](https://github.com/metabase/metabase/blob/9ca4af5b3c9c18956e7c0d545fb029d21e7c452e/frontend/src/metabase/query_builder/components/__mocks__/NativeQueryEditor.jsx) would have been used in tests instead of the actual implementation.

ranquild reviewed (commented):
Some parts are unclear to me, including one about the async import.

kamilmielnik committed:
Use createMockCollection instead of a cast

kamilmielnik committed:
Move all Native Query creation to createMockNativeDatasetQuery

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,152 @@
+
+describe("DatasetQueryEditor", () => {
+  beforeEach(() => {
+    fetchMock.get("path:/api/collection", [ROOT_COLLECTION]);
```
`setupCollectionEndpoints`

kamilmielnik replied:
c31950053c5609e46525443a9bc92da29f205520

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,152 @@
+describe("DatasetQueryEditor", () => {
+  beforeEach(() => {
+    fetchMock.get("path:/api/collection", [ROOT_COLLECTION]);
+    fetchMock.get("path:/api/native-query-snippet", () => []);
```
This one might not yet exist in `server-mocks`, if that's the case, can you add it there? we use `clj` api endpoint file names for the file names in the mocks folder.

kamilmielnik replied:
9c7f57c81cb2a4095c7ae9b4fce5c4636c1dad56

ranquild commented on frontend/src/metabase/query_builder/components/DatasetEditor/DatasetQueryEditor.unit.spec.tsx:
```diff
@@ -0,0 +1,152 @@
+    fetchMock.get("path:/api/collection", [ROOT_COLLECTION]);
+    fetchMock.get("path:/api/native-query-snippet", () => []);
+
+    jest.unmock("metabase/query_builder/components/NativeQueryEditor");
```
Could we maybe extract this into a function so we can put the async import with a comment about the mocked component there? And use this function in 2 places in the test suite.

ranquild replied:
If not, that's ok

kamilmielnik replied:
a74fea711fc8910bf7739178f313d95d2baa7bac

ranquild reviewed (commented):
Last comments about using `server-mocks`

kamilmielnik committed:
Use setupCollectionsEndpoints

kamilmielnik committed:
Introduce setupNativeQuerySnippetEndpoints and use it in DatasetQueryEditor.unit.spec.tsx

kamilmielnik committed:
Extract explanatory importDatasetQueryEditor

kamilmielnik committed:
Merge branch 'master' into 30680-native-editor-buttons

kamilmielnik committed:
Remove file accidentally added when merging master

kamilmielnik committed:
Add missing data-testid that got lost during a merge

kamilmielnik committed:
Merge branch 'master' into 30680-native-editor-buttons

kamilmielnik committed:
Merge branch 'master' into 30680-native-editor-buttons
